src/software/sensor_fusion/filter/ball_filter.h

```diff
@@ -52,25 +45,67 @@ class BallFilter
-     *
-     * @param min_buffer_size The minimum size of the buffer the filter will use to filter
-     * the ball. The buffer will shrink to this size as the ball speeds up
-     * @param min_buffer_size The maximum size of the buffer the filter will use to filter
-     * the ball. The buffer will grow to this size as the ball slows down
      */
-    explicit BallFilter(unsigned int min_buffer_size = DEFAULT_MIN_BUFFER_SIZE,
-                        unsigned int max_buffer_size = DEFAULT_MAX_BUFFER_SIZE);
+    explicit BallFilter();
 
     /**
      * Filters the new ball detection data, and returns the updated state of the ball
-     * given the new data *
-     * @param current_ball_state The current state of the Ball
+     * given the new data
+     *
      * @param new_ball_detections A list of new Ball detections
+     * @param filter_area The area within which the ball filter will work. Any detections
+     * outside of this area will be ignored.
      *
      * @return The updated state of the ball given the new data
      */
-    std::optional<TimestampedBallState> getFilteredData(
-        const std::vector<BallDetection> &new_ball_detections, const Field &field);
+    std::optional<TimestampedBallState> estimateBallState(
+        const std::vector<BallDetection>& new_ball_detections,
+        const Rectangle& filter_area);
+
+   private:
+    /**
+     * A simple struct we use to pass around velocity estimate data
+     */
+    struct BallVelocityEstimate
+    {
+        Vector average_velocity;
+        double average_velocity_magnitude;
+        // The average of the max velocity magnitude and min velocity magnitude
+        double min_max_magnitude_average;
+    };
+
+    /**
+     * A simple struct to pass around linear regression data
+     */
+    struct LinearRegressionResults
+    {
+        Line regression_line;
+        double regression_error;
+    };
+
+    /**
+     * Adds ball detections to the buffer stored by this filter. This function will ignore
+     * data that is outside of the filter_area, and data that is too far away from the
+     * current known ball position (since it is likely to be random noise).
+     *
+     * @param new_ball_detections The ball detections to try add to the buffer
+     * @param filter_area The area within which the ball filter will work. Any detections
+     * outside of this area will be ignored.
+     */
+    void addNewDetectionsToBuffer(std::vector<BallDetection> new_ball_detections,
+                                  const Rectangle& filter_area);
+
+    /**
+     * Uses linear regression to filter the given list of ball detections to find the
+     * current "real" state of the ball.
+     *
+     * @param ball_detections The detections to filter
+     *
+     * @return The filtered current state of the ball. If a filtered result cannot be
+     * calculated, returns std::nullopt
+     */
+    static std::optional<TimestampedBallState> estimateBallState(
```

Comment:
Can this be an overload of the other estimateBallState?

Reply:
Is that preferable? They're pretty close as-is (and can't be an overload if one is `static` I think). I'm not sure I totally understand what you're going for here?

Reply:
it gets confusing when functions with the same name do subtly different things, namely `addNewDetectionsToBuffer`

Reply:
I'm still confused. How does an overload fix that? That would still be functions with the same name?

Reply:
changed to `estimateBallStateFromBuffer` to disambiguate